### PR TITLE
Support lists in expressions

### PR DESCRIPTION
**Pull request by TomasEng**
## Description
This pull request adds support for arrays, along with a `list` function, in expressions.

- `list` function is added
- Array support is added to `ExpressionValue` (most of the code was already there, hidden in comments)
- Made the tests support test files with several test cases and undefined root expression (currently, this is only supported for functions with a simple test file, but that is not the case for `dataModel`, of which the new file `lookup-list.json` contain several test cases)

Here is the corresponding change in App Frontend: https://github.com/Altinn/altinn-studio/pull/18974

## Related Issue(s)
- https://github.com/Altinn/app-frontend-react/issues/4189

## Verification
- [x] **Your** code builds clean without any errors or warnings
- Not possible to test manually until we have added aggregate functions (without which there is no use case for lists), but the tests should be sufficient
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs): https://github.com/Altinn/altinn-studio-docs/pull/2894

## Summary by CodeRabbit

* **New Features**
  * Added list expression support so expressions can create, return and evaluate array/list values.

* **Tests**
  * Added and expanded fixtures and unit tests covering list creation, empty/nested/mixed lists, lookups, parsing/serialization, evaluation behavior, and related edge cases.

### CODE REVIEW

**Comment by TomasEng** [comment]
@coderabbitai review

File: src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs [hunk 1]
```diff
@@ -89,11 +90,12 @@ private ExpressionValue(string? value)
...
-    // {
-    //     _valueKind = value is null ? JsonValueKind.Null : JsonValueKind.Array;
-    //     _arrayValue = value;
-    // }
+    /// <summary>Constructor for array value</summary>
+    public ExpressionValue(ExpressionValue[] value)
+    {
+        ValueKind = JsonValueKind.Array;
+        _arrayValue = value;
+    }
```
[the remark this answers is not in the document]
Comment (TomasEng) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs]: What is the purpose of these checks? The typing makes it clear that the value is always an array and never null.

**Comment by TomasEng** [comment]
@coderabbitai
> Deserialize array directly to avoid an extra allocation.

Fixed.

**Comment by TomasEng** [comment]
The Sonar errors on this pull request seem to be related to existing code only.

File: test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/TestBackendExclusiveFunctions.cs [hunk 2]
```diff
@@ -93,7 +94,7 @@ private async Task RunTestCase(string testName, string folder)
                 {
                     await ExpressionEvaluator.EvaluateExpression(
                         state,
-                        test.Expression,
+                        (Expression)test.Expression!,
```
Comment (ivarne) [inline test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/TestBackendExclusiveFunctions.cs]: I think these should be `test.Expression.Value`

Comment (TomasEng) [reply test/Altinn.App.Core.Tests/LayoutExpressions/CommonTests/TestBackendExclusiveFunctions.cs]: Right, fixed!

File: src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs [hunk 3]
```diff
@@ -997,6 +999,11 @@ private static ExpressionValue Argv(ExpressionValue[] args, ExpressionValue[]? p
         return positionalArguments[index.Value];
     }
 
+    private static JsonArray List(ExpressionValue[] args)
```
Comment (ivarne) [inline src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs]: La denne returnere `ExpressionValue` i stedet for å la en implisit cast gjøre det etterpå

File: src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs [hunk 4]
```diff
@@ -172,6 +174,17 @@ public static ExpressionValue FromObject(object? value)
                     '"'
                 ) // Trim quotes to match the string representation
             ,
+            JsonArray jsonArrayValue => jsonArrayValue,
+            _ => ToJsonArrayOrNull(value),
+        };
+    }
+
+    private static ExpressionValue ToJsonArrayOrNull(object? value)
```
Comment (ivarne) [inline src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs:182]: Dette er forsåvidt en fornuftig fallback, men navngivningen er litt rar, synes jeg, og hvis vi først trenger en fallback som bruker JsonNode, burde den vel kunne laget `ExpressionValue` av noder som kommer tilbake som både "string", "number", "bool" og alt annet.

Jeg foreslår å navngi den `FromObjectUsingNodeSerializer` og håndtere alle json node typer i switch expression nederst.

**Review by ivarne — APPROVED** [review]
God start!

Bare noen ubetydelige kommentarer på ting jeg ikke umiddelbart forsto at var grei kode.

File: src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs (hunk 4 above)
Comment (ivarne) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionValue.cs:182]: Stryk det, ser at den endrer navn i neste PR

File: src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs (hunk 3 above)
Comment (TomasEng) [reply src/Altinn.App.Core/Internal/Expressions/ExpressionEvaluator.cs]: Fikset

**Comment by TomasEng** [comment]
Fjernet en test som testet en umulig testcase. Se https://github.com/Altinn/altinn-studio/pull/18974#discussion_r3419233862.
